### PR TITLE
Problem: coin overflow check (CRO-168)

### PR DESCRIPTION
**Pull request by leejw51crypto**
Solution: add unit test using quickcheck

### CODE REVIEW

**Comment by leejw51crypto** [comment]
add unit tests

File: chain-core/src/init/coin.rs [hunk 1]
```diff
@@ -241,6 +241,39 @@ mod test {
     use quickcheck::quickcheck;
 
     quickcheck! {
+        // test whether oveflow error occur
```
Comment (tomtau) [inline chain-core/src/init/coin.rs]: Quickcheck is meant for property tests where the argument to the test functions are fed with random inputs, so it's a bit confusing / redundant to have unit tests here.

it'll be better to move these up (from the quickcheck macro) as regular units in Rust:
```
    #[test]
    fn coin_overflow_add() {
    ...
    assert!(sum.is_err())
    }
```

and perhaps give the more meaningful names than `coin_add`, `coin_add2`...

For example: `add_2_coin_max_should_produce_error`

File: chain-core/src/init/coin.rs [hunk 2]
```diff
@@ -241,6 +241,39 @@ mod test {
...
+            let a = Coin::max();
+            let b = Coin::max();
+            let sum = a+ b;
+            sum.is_err()
+        }
+
+        // test whether overflow error not occur
+          fn coin_add2() -> bool {
+            let a = Coin::max();
+            let b = Coin::new(0).unwrap();
```
Comment (tomtau) [inline chain-core/src/init/coin.rs]: could this actually cause overflows?
The "lowest" overflowing addition is: Coin::max() + Coin::new(84_467_440_737__0955_1616), I think?

File: chain-core/src/init/coin.rs [hunk 3]
```diff
@@ -241,6 +241,39 @@ mod test {
...
+            let a = Coin::new(0).unwrap();
+            let b = Coin::max();
+            let sub = a- b;
+            sub.is_err()
+        }
+
+        // test whether underflow error not occur
+        fn coin_sub2() -> bool {
+            let a = Coin::max();
+            let b = Coin::new(0).unwrap();
```
Comment (tomtau) [inline chain-core/src/init/coin.rs]: could this cause underflow?

File: chain-core/src/init/coin.rs (hunk 2 above)
Comment (leejw51crypto) [reply chain-core/src/init/coin.rs]: this is not overflow.
this is assert for "max + 0" works fine

File: chain-core/src/init/coin.rs (hunk 1 above)
Comment (leejw51crypto) [reply chain-core/src/init/coin.rs]: done

File: chain-core/src/init/coin.rs (hunk 3 above)
Comment (leejw51crypto) [reply chain-core/src/init/coin.rs]: no

File: chain-core/src/init/coin.rs (hunk 2 above)
Comment (tomtau) [reply chain-core/src/init/coin.rs]: ok, you may want to add a test case with `Coin::new(84_467_440_737__0955_1616)`

File: chain-core/src/init/coin.rs (hunk 3 above)
Comment (tomtau) [reply chain-core/src/init/coin.rs]: so perhaps this test is not necessary?

**Pull request merged**